File: cpp/lru-cache.cpp
class LRUCache {
public:
    LRUCache(int capacity)
     : size{capacity} { }

    int get(int key) {
        // Return -1 if the key was not in the map
        if (hashmap.count(key) == 0) {
            return -1;
        }
        update_cache(key);
        return hashmap[key];
    }
    
    
    void put(int key, int value) {
        if (hashmap.size() == size and hashmap.count(key) == 0) {
            clear_cache();
        }
        update_cache(key);
        hashmap[key] = value;
    }

private:
    int size;
    list<int> lru;
    unordered_map<int, list<int>::iterator> map_itr;
    unordered_map<int, int> hashmap;

    // clears the lru cache
    void clear_cache() {
        map_itr.erase(lru.back());
        hashmap.erase(lru.back());
        lru.pop_back();
    }

    // adds a new key to the cache
    void update_cache(int key) {
        if (hashmap.count(key) == 1) {
            lru.erase(map_itr[key]);
        }

        lru.push_front(key); // O(1)

        map_itr[key] = lru.begin();
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: py/lru-cache.py
class Node:
    def __init__(self, k, v: int):
        ''' initiliazation of a Node for LRUCache '''
        # initialize the key and value of the node
        self.key = k
        self.val = v

        # pointers for the doubly-linked list
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, cap: int):
        self.capacity = cap
        self.dic = {}

        # linking the head/tail pointers
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head

    def get(self, key: int) -> int:
        '''
        get() returns the value (will always be positive) of the key if
        the key exists in the cache, otherwise it returns -1.
        '''
        if key in self.dic:
            n = self.dic[key]
            self._remove(n)
            self._add(n)
            return n.val
        return -1

    def put(self, k: int, val: int) -> None:
        '''
        put() - Set or insert the value if the key is not already
        present. When the cache reached its capacity, it should invalidate the
        least recently used item before inserting a new item.
        '''
        if k in self.dic:
            self._remove(self.dic[k])
        n = Node(k, val)
        self._add(n)
        self.dic[k] = n

        # check to see if we go over capacity
        if len(self.dic) > self.capacity:
            n = self.head.next
            self._remove(n)
            del self.dic[n.key]


    def _add(self, node: Node) -> None:
        '''
        helper function to rearrange pointers in the doubly-linked list when we
        add a Node to the cache
        '''
        p = self.tail.prev
        p.next = node
        self.tail.prev = node
        node.prev = p
        node.next = self.tail


    def _remove(self, node: Node) -> None:
        '''
        helper function to rearrange pointers in the doubly-linked list when we
        remove a Node from the cache
        '''
        p = node.prev
        n = node.next
        p.next = n
        n.prev = p




# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)